### PR TITLE
feat: [CDS-76887]: Support fetching by commit SHA in Git Clone step

### PR DESCRIPTION
**Pull request by tmacari-harness**
## Describe your changes
feat: [CDS-76887]: Support fetching by commit SHA in Git Clone step

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

[CDS-76887]: https://harness.atlassian.net/browse/CDS-76887?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51686)

### CODE REVIEW

**Comment by tmacari-harness** [comment]
trigger  uts

**Comment by tmacari-harness** [comment]
trigger uts

File: 953-yaml-commons/src/main/java/io/harness/yaml/extended/ci/codebase/GitCloneStepBuild.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.List;
+import javax.validation.constraints.NotNull;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class GitCloneStepBuild {
```
Comment (raghavharness) [inline 953-yaml-commons/src/main/java/io/harness/yaml/extended/ci/codebase/GitCloneStepBuild.java:22]: Any particular reason to create a new class here ? Can't we use the existing `Build` class ?

File: 879-pipeline-ci-commons/src/main/java/io/harness/plugin/service/PluginServiceImpl.java [hunk 2]
```diff
@@ -282,4 +290,12 @@ static Map<String, String> getPluginOutputFilePathsContent(
   public Map<String, SecretNGVariable> getPluginCompatibleSecretVars(PluginCompatibleStep step, String identifier) {
     return new HashMap<>();
   }
+
+  private static GitCloneStepBuild resolveBuild(ParameterField<GitCloneStepBuild> buildDetails) {
```
Comment (harshjain12) [inline 879-pipeline-ci-commons/src/main/java/io/harness/plugin/service/PluginServiceImpl.java:294]: Move it to io.harness.beans.serializer.RunTimeInputHandler

**Comment by harshjain12** [comment]
1. Move logic to Runtime handler
2. Backward compatibility.
3. Check yaml for codebase (It should not contain commitsha).

File: 879-pipeline-ci-commons/src/main/java/io/harness/beans/steps/stepinfo/GitCloneStepInfo.java [hunk 3]
```diff
@@ -103,10 +103,10 @@ public class GitCloneStepInfo implements PluginCompatibleStep, WithConnectorRef
 
   @Builder
   @ConstructorProperties({"identifier", "name", "retry", "connectorRef", "resources", "runAsUser", "repoName", "build",
-      "projectName", "depth", "sslVerify", "cloneDirectory", "outputFilePathsContent"})
+      "projectName", "depth", "sslVerify", "test", "cloneDirectory", "outputFilePathsContent"})
   public GitCloneStepInfo(String identifier, String name, int retry, ParameterField<String> connectorRef,
       ContainerResource resources, ParameterField<Integer> runAsUser, ParameterField<String> repoName,
-      ParameterField<Build> build, ParameterField<String> projectName, ParameterField<Integer> depth,
+      ParameterField<GitCloneStepBuild> build, ParameterField<String> projectName, ParameterField<Integer> depth,
```
Comment (sahilhindwani-harness) [inline 879-pipeline-ci-commons/src/main/java/io/harness/beans/steps/stepinfo/GitCloneStepInfo.java:109]: Discussed offline, this change is backward incompatible.
All running executions will try to use Build field while the new code will use GitCloneStepBuild resulting in ClasscastException.

We should use Build only and add new type over there.

**Pull request closed without merging**